### PR TITLE
Process data implementation

### PR DESCRIPTION
**Pull request by KKarlash**
@bonewell 
@DKlimenko 
@VladSemenyuk 
@thinhts 
@giangtqh

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ */
+class ProcessDataRequest : public BaseCommandRequest {
+ public:
+  /**
+   * @brief ProcessDataRequest class constructor
+   * @param parent pointer to VRModule
+   * @param message Message from HMI
+   */
+  ProcessDataRequest(VRModule* parent,
+                         const vr_hmi_api::ServiceMessage& message);
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: indent

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_PROCESS_DATA_REQUEST_H_
+#define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_PROCESS_DATA_REQUEST_H_
+
+#include "vr_cooperation/commands/base_command_request.h"
+#include "vr_cooperation/vr_module.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: Please, use forward declaration.

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: include process_data_request.h

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+std::string ProcessDataRequest::GetParams(
+    const Json::Value& value) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  vr_hmi_api::ResultCode result_code;
+  ParseMobileResultCode(value, result_code);
+  vr_hmi_api::ProcessDataResponse response;
+  response.set_result(result_code);
+  response.set_allocated_text(value[kText]);
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Strange! Why not `response.set_text(value[kText].asString());` ?

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "functional_module/function_ids.h"
+#include "vr_cooperation/commands/activate_service_request.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Why?

**Comment by VladSemenyuk** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h (hunk 2 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: There is information about ServiceMessage

Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: Use forward declaration for ServiceMessage too.
```
namespace vr_hmi_api {
class ServiceMessage;
}  // namespace vr_hmi_api
```

**Comment by VladSemenyuk** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace vr_cooperation {
+
+namespace commands {
+
+/**
+ * @brief ProcessDataRequest command class
+ */
+class ProcessDataRequest : public BaseCommandRequest {
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: I think it should inherit from RequestFromHMI

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      const event_engine::Event<application_manager::MessagePtr,
+                                vr_hmi_api::RPCName>& event);
+
+ private:
+  /**
+   * @brief Gets params from json value
+   * @param value json value
+   * @returns strings with params for GPB message
+   */
+  std::string GetParams(const Json::Value& value);
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: const function

File: src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+  /**
+   * @brief Prepares GPB message for HMI
+   * @param message GPB message for HMI
+   * @param value from json message
+   */
+  void PrepareGpbMessage(const Json::Value& value,
+                         vr_hmi_api::ServiceMessage& message);
```
Comment (thinhts) [inline src/components/vr_cooperation/include/vr_cooperation/commands/process_data_request.h]: const function

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 9]
```diff
@@ -30,19 +30,23 @@
...
+using json_keys::kInfo;
+
 CREATE_LOGGERPTR_GLOBAL(logger_, "ProcessDataRequest")
 
 ProcessDataRequest::ProcessDataRequest(
-      VRModule* parent,
-      const vr_hmi_api::ServiceMessage& message)
+    VRModule* parent,
+    const vr_hmi_api::ServiceMessage& message)
   : BaseCommandRequest(parent, message) {
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Wrong indent.

Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Wrong indent.

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 10]
```diff
@@ -85,7 +88,7 @@ std::string ProcessDataRequest::GetParams(
 
 void ProcessDataRequest::PrepareGpbMessage(
     const Json::Value& value,
-    vr_hmi_api::ServiceMessage& message) {
+    vr_hmi_api::ServiceMessage& message){
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Space between `)` and `{`

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/src/process_data_request.cc [hunk 11]
```diff
@@ -45,8 +45,8 @@ using json_keys::kInfo;
 CREATE_LOGGERPTR_GLOBAL(logger_, "ProcessDataRequest")
 
 ProcessDataRequest::ProcessDataRequest(
-    VRModule* parent,
-    const vr_hmi_api::ServiceMessage& message)
+      VRModule* parent,
+      const vr_hmi_api::ServiceMessage& message)
   : BaseCommandRequest(parent, message) {
```
Comment (bonewell) [inline src/components/vr_cooperation/src/process_data_request.cc]: Wrong indent of 3 lines now. You had wrong indent in 50 line: [Constructor_Initializer_Lists](https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists)

**Comment by bonewell** [comment]
Reviewed.

**Comment by DKlimenko** [comment]
reviewed

**Comment by giangtqh** [comment]
reviewed

**Pull request merged**